pkg/util/apiserver/apiserver_test.go

```diff
@@ -0,0 +1,220 @@
+	ts := httptest.NewUnstartedServer(mux)
+	api := NewAPIServer(mux)
+
+	api.AddAPIResource(gv.String(), resource, func(_ http.ResponseWriter, _ *http.Request, _ string) error {
+		return nil
+	})
+
+	ts.Start()
+	defer ts.Close()
+
+	client := ts.Client()
+	path := ts.URL + "/apis/allocation.agones.dev/v1alpha1"
+
+	t.Run("No Accept Header", func(t *testing.T) {
+		resp, err := client.Get(path)
+		if resp != nil {
+			defer resp.Body.Close() // nolint: errcheck
+		}
+		if !assert.NoError(t, err) {
+			assert.FailNow(t, "should not error")
+		}
+		assert.Equal(t, k8sruntime.ContentTypeJSON, resp.Header.Get("Content-Type"))
+
+		// default is json
+		list := &metav1.APIResourceList{}
+		err = json.NewDecoder(resp.Body).Decode(list)
+		assert.NoError(t, err)
+
+		assert.Equal(t, "v1", list.TypeMeta.APIVersion)
+		assert.Equal(t, "APIResourceList", list.TypeMeta.Kind)
+		assert.Equal(t, gv.String(), list.GroupVersion)
+		assert.Equal(t, resource, list.APIResources[0])
+	})
+
+	t.Run("Accept */*", func(t *testing.T) {
+		request, err := http.NewRequest(http.MethodGet, path, nil)
+		assert.NoError(t, err)
+
+		request.Header.Set("Accept", "*/*")
+		resp, err := client.Do(request)
+		assert.NoError(t, err)
+		if resp != nil {
+			defer resp.Body.Close() // nolint: errcheck
+		}
+		assert.Equal(t, k8sruntime.ContentTypeJSON, resp.Header.Get("Content-Type"))
+
+		list := &metav1.APIResourceList{}
+		err = json.NewDecoder(resp.Body).Decode(list)
+		assert.NoError(t, err)
+
+		assert.Equal(t, "v1", list.TypeMeta.APIVersion)
+		assert.Equal(t, "APIResourceList", list.TypeMeta.Kind)
+		assert.Equal(t, gv.String(), list.GroupVersion)
+		assert.Equal(t, resource, list.APIResources[0])
+	})
+
+	t.Run("Accept Protobuf, */*", func(t *testing.T) {
+		request, err := http.NewRequest(http.MethodGet, path, nil)
+		assert.NoError(t, err)
+
+		request.Header.Set("Accept", "application/vnd.kubernetes.protobuf, */*")
+		resp, err := client.Do(request)
+		assert.NoError(t, err)
+		if resp != nil {
+			defer resp.Body.Close() // nolint: errcheck
+		}
+		assert.Equal(t, "application/vnd.kubernetes.protobuf", resp.Header.Get("Content-Type"))
+
+		list := &metav1.APIResourceList{}
+		b, err := ioutil.ReadAll(resp.Body)
+		assert.NoError(t, err)
+
+		info, ok := k8sruntime.SerializerInfoForMediaType(Codecs.SupportedMediaTypes(), "application/vnd.kubernetes.protobuf")
+		assert.True(t, ok)
+
+		gvk := unversionedVersion.WithKind("APIResourceList")
+		_, _, err = info.Serializer.Decode(b, &gvk, list)
+		assert.NoError(t, err)
+
+		// not testing for version/kind, seems not to come through. Strange.
+		// we're not building anything that needs this, so not our issue.
+		assert.Equal(t, gv.String(), list.GroupVersion)
+		assert.Equal(t, resource, list.APIResources[0])
+	})
+}
+
+func TestSplitNameSpaceResource(t *testing.T) {
+	type expected struct {
+		namespace string
+		resource  string
+		isError   bool
+	}
+
+	fixtures := []struct {
+		path     string
+		expected expected
```

Comment:
you could use an anonymous struct.

Reply:
AFAIK you can't nest anonymous structs, so I've always used this approach.

Reply:
ho yes definitively.